planner/core/rule_generate_column_substitute.go

```diff
@@ -97,26 +100,38 @@ func appendSubstituteColumnStep(lp LogicalPlan, candidateExpr expression.Express
 	opt.appendStepToCurrent(lp.ID(), lp.TP(), reason, action)
 }
 
-func substituteExpression(cond expression.Expression, lp LogicalPlan, exprToColumn ExprColumnMap, schema *expression.Schema, opt *logicalOptimizeOp) {
+// SubstituteExpression is Exported for bench
+func SubstituteExpression(cond expression.Expression, lp LogicalPlan, exprToColumn ExprColumnMap, schema *expression.Schema, opt *logicalOptimizeOp) bool {
+	return substituteExpression(cond, lp, exprToColumn, schema, opt)
+}
+
+func substituteExpression(cond expression.Expression, lp LogicalPlan, exprToColumn ExprColumnMap, schema *expression.Schema, opt *logicalOptimizeOp) bool {
 	sf, ok := cond.(*expression.ScalarFunction)
 	if !ok {
-		return
+		return false
 	}
 	sctx := lp.SCtx().GetSessionVars().StmtCtx
+	changed := false
+	collectChanged := func(partial bool) {
+		if partial && !changed {
+			changed = true
+		}
+	}
 	defer func() {
 		// If the argument is not changed, hash code doesn't need to recount again.
-		// But we always do it to keep the code simple and stupid.
-		expression.ReHashCode(sf, sctx)
+		if changed {
+			expression.ReHashCode(sf, sctx)
```

Comment:
for better we could even just clean the hashcode as [:0] here because the computed hashcode or top  OR expression is never being used.